Wikipedia/Code/WelcomeContainerViewController.swift

```diff
@@ -0,0 +1,71 @@
+import UIKit
+
+protocol WelcomeContainerViewControllerDataSource: AnyObject {
+    var foregroundAnimationView: WelcomeAnimationView { get }
+    var backgroundAnimationView: WelcomeAnimationView? { get }
+    var panelViewController: WelcomePanelViewController { get }
+}
+
+class WelcomeContainerViewController: UIViewController {
+    weak var dataSource: WelcomeContainerViewControllerDataSource?
+    
+    @IBOutlet private weak var topForegroundContainerView: UIView!
+    @IBOutlet private weak var topBackgroundContainerView: UIView!
+    @IBOutlet private weak var bottomContainerView: UIView!
+
+    private var theme = Theme.standard
+
+    init(dataSource: WelcomeContainerViewControllerDataSource) {
+        self.dataSource = dataSource
+        super.init(nibName: "WelcomeContainerViewController", bundle: Bundle.main)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        guard let dataSource = dataSource else {
+            assertionFailure("dataSource should be set by now")
+            apply(theme: theme)
+            return
+        }
+        addChild(WelcomeAnimationViewController(position: .foreground, animationView: dataSource.foregroundAnimationView), to: topForegroundContainerView)
+        if let backgroundAnimationView = dataSource.backgroundAnimationView {
+            addChild(WelcomeAnimationViewController(position: .background, animationView: backgroundAnimationView), to: topBackgroundContainerView)
+        } else {
+            topBackgroundContainerView.isHidden = true
+        }
+        addChild(dataSource.panelViewController, to: bottomContainerView)
+        apply(theme: theme)
+    }
+    
+    private func addChild(_ viewController: UIViewController?, to view: UIView) {
```

Comment:
I notice there's a few similar chunks of this code, we could move adding a child VC to a UIViewController extension at some point

Reply:
we have `@objc public func wmf_add(childController: UIViewController!, andConstrainToEdgesOfContainerView containerView: UIView!)` that I forgot about 🙈